Sources/SMTPKitten/Connection/SMTPClient.swift

```diff
@@ -0,0 +1,157 @@
+                                code: code,
+                                lines: lines.map(\.contents)
+                            )
+                        )
+                    } catch {
+                        request.continuation.resume(throwing: error)
+                        throw error
+                    }
+                }
+            }
+
+            requestWriter.finish()
+            for await request in requests {
+                request.continuation.resume(throwing: SMTPClientError.endOfStream)
+            }
+        } catch {
+            self.error = error
+            requestWriter.finish()
+            for await request in requests {
+                request.continuation.resume(throwing: error)
+            }
+            throw error
+        }
+    }
+
+    nonisolated fileprivate func starttls(
+        configuration: SMTPSSLConfiguration,
+        hostname: String
+    ) async throws {
+        try await send(.starttls)
+            .status(.serviceReady, or: SMTPClientError.startTLSFailure)
+
+        let sslContext = try NIOSSLContext(configuration: configuration.configuration.makeTlsConfiguration())
+        let sslHandler = try NIOSSLClientHandler(context: sslContext, serverHostname: hostname)
+
+        try await self.channel.channel.pipeline.addHandler(sslHandler, position: .first)
+    }
+
+    public static func withConnection<T>(
+        to host: String,
+        port: Int = 587,
+        ssl: SMTPSSLMode,
+        perform: (SMTPClient) async throws -> T
+    ) async throws -> T {
+        let asyncChannel: NIOAsyncChannel<SMTPReplyLine, ByteBuffer> = try await ClientBootstrap(
+            group: NIOSingletons.posixEventLoopGroup
+        ).connect(host: host, port: port) { channel in
+            do {
+                if case .tls(let tls) = ssl.mode {
+                    let context = try NIOSSLContext(
+                        configuration: tls.configuration.makeTlsConfiguration()
+                    )
+
+                    try channel.pipeline.syncOperations.addHandler(
+                        NIOSSLClientHandler(context: context, serverHostname: host)
+                    )
+                }
+
+                try channel.pipeline.syncOperations.addHandlers(
+                    ByteToMessageHandler(LineBasedFrameDecoder()),
+                    ByteToMessageHandler(SMTPReplyDecoder())
+                )
+
+                let asyncChannel = try NIOAsyncChannel<SMTPReplyLine, ByteBuffer>(synchronouslyWrapping: channel)
+                return channel.eventLoop.makeSucceededFuture(asyncChannel)
+            } catch {
+                return channel.eventLoop.makeFailedFuture(error)
+            }
+        }
+
+        let client = SMTPClient(channel: asyncChannel)
+        return try await withThrowingTaskGroup(of: Void.self) { group in
+            group.addTask {
+                try await client.run()
+            }
+
+            let requestWriter = client.requestWriter
```

Comment:
Could you not use `send()` here. Also a comment indicating why you are sending an empty message might help, although you might not want to use `send()` as it should probably protect against sending empty `ByteBuffers`, because that'll confuse the loop in `run()`